rfcs/text/0007_lifecycle_unblocked.md

```diff
@@ -0,0 +1,339 @@
+2. This opens up the potential for a bug in Elastic or third-party plugins to
+   effectively "break" kibana. Instead of a single plugin being disabled all
+   of kibana would be down requiring manual intervention by a system
+   administrator. 
+
+# Adoption strategy
+Adoption and implementation should be handled as follows:
+
+1. Don't expose core API's as observables (3).
+   This should be implemented first to improve the ergonomics of working with
+   core API's from inside synchronous context providers and lifecycle functions.
+   
+   Plugins consuming observable-based API's generally follow a pattern like the
+   following which effectively ignores the observable and should be easy to
+   refactor:
+
+   ```diff
+   - const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(take(1)).toPromise();
+   + const adminClient = coreSetup.elasticsearch.adminClient;
+   ```
+
+2. Making context provider functions synchronous (2)
+   Adoption of context provider functions is still fairly low, so the amount
+   of change required by plugin authors should be limited.
+
+3. Synchronous lifecycle methods (1) will have the biggest impact on plugins
+   since many NP plugins and shims have been built with async lifecycle
+   methods in mind.
+
+   The following New Platform plugins or shims currently rely on async lifecycle functions:
+   1. [region_map](https://github.com/elastic/kibana/blob/6039709929caf0090a4130b8235f3a53bd04ed84/src/legacy/core_plugins/region_map/public/plugin.ts#L68)
+   2. [tile_map](https://github.com/elastic/kibana/blob/6039709929caf0090a4130b8235f3a53bd04ed84/src/legacy/core_plugins/tile_map/public/plugin.ts#L62)
+   3. [vis_type_table](https://github.com/elastic/kibana/blob/6039709929caf0090a4130b8235f3a53bd04ed84/src/legacy/core_plugins/vis_type_table/public/plugin.ts#L61)
+   4. [vis_type_vega](https://github.com/elastic/kibana/blob/6039709929caf0090a4130b8235f3a53bd04ed84/src/legacy/core_plugins/vis_type_vega/public/plugin.ts#L59)
+   5. [timelion](https://github.com/elastic/kibana/blob/9d69b72a5f200e58220231035b19da852fc6b0a5/src/plugins/timelion/server/plugin.ts#L40)
+   6. [code](https://github.com/elastic/kibana/blob/5049b460b47d4ae3432e1d9219263bb4be441392/x-pack/legacy/plugins/code/server/plugin.ts#L129-L149)
+   7. [spaces](https://github.com/elastic/kibana/blob/096c7ee51136327f778845c636d7c4f1188e5db2/x-pack/legacy/plugins/spaces/server/new_platform/plugin.ts#L95)
+   8. [licensing](https://github.com/elastic/kibana/blob/4667c46caef26f8f47714504879197708debae32/x-pack/plugins/licensing/server/plugin.ts)
+   9. [security](https://github.com/elastic/kibana/blob/0f2324e44566ce2cf083d89082841e57d2db6ef6/x-pack/plugins/security/server/plugin.ts#L96)
+
+Once this RFC has been merged we should educate teams on the implications to
+allow existing New Platform plugins to remove any async lifecycle
+behaviour they're relying on before we change this in core.
```

Comment:
Migrating to sync lifecycle methods only affects API's which required an async operation to resolve before the API was returned from `setup()`. This means that some API's might change, but if plugin A hasn't been migrated it wouldn't block an upstream plugin from migrating. It will however create more work later on. So even if Plugin B has finished migrating, once Plugin A changes it's API Plugin B will once again have to do some refactoring to adopt the new API.

```ts
// Dependency with unmigrated, async lifecycles
class PluginA {
  async setup(core: CoreSetup) {
    return {
      create: async () => {return await core.savedObjects.getInternalRepository.create(...)}
    }
  }
}

// Migrated plugin with sync lifecycles
class PluginB {
  setup(core: CoreSetup, deps) {
    // const wasCreateSuccessful = (await deps.pluginA.create()) != null;
    deps.pluginA.create().then(() => {
      ... // I can consume an async API in setup, and wait till it resolves before
      ... // doing further operations, I just can't block setup.
    });
    return {
      // Plugins can continue to expose API's that make further async calls
      // to dependencies
      proxiedCreate: async () => {return await deps.pluginA.create(...)} 
      /* wasCreateSuccessful it's now impossible to expose this API */ 
    }
  }
}